src/utils/concurrent_container.h

```diff
@@ -14,6 +14,9 @@
 #define READER_LOCK(mu) std::shared_lock<std::shared_mutex> reader(mu);
 #define WRITER_LOCK(mu) std::unique_lock<std::shared_mutex> writer(mu);
 
+auto key_selector   = [](const auto& pair) { return pair.first; };
```

Comment:
why it is moved from private to outside?

Reply:
Because it’s paired with the `value_selector`

Reply:
I would recommend not moving them out but writing this simple lambda function explicitly when using for encapsulation and readability.